includes/SVG.php

```diff
@@ -0,0 +1,428 @@
+	 * Speeds up access during the same request.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @var string[]
+	 */
+	protected $svgs = [];
+
+	/**
+	 * Experiments instance.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @var Experiments Experiments instance.
+	 */
+	private $experiments;
+
+	/**
+	 * SVG constructor.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param Experiments $experiments Experiments instance.
+	 *
+	 * @return void
+	 */
+	public function __construct( Experiments $experiments ) {
+		$this->experiments = $experiments;
+	}
+
+	/**
+	 * Register filters and actions.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @return void
+	 */
+	public function init() {
+		if ( ! $this->experiments->is_experiment_enabled( 'enableSVG' ) ) {
+			return;
+		}
+
+		add_filter( 'web_stories_allowed_mime_types', [ $this, 'web_stories_allowed_mime_types' ] );
+
+		// Check if svg uploads, already enabled.
+		if ( $this->svg_already_enabled() ) {
+			return;
+		}
+
+		add_filter( 'upload_mimes', [ $this, 'upload_mimes_add_svg' ] ); // phpcs:ignore WordPressVIPMinimum.Hooks.RestrictedHooks.upload_mimes
+		add_filter( 'mime_types', [ $this, 'mime_types_add_svg' ] );
+		add_filter( 'wp_handle_upload_prefilter', [ $this, 'wp_handle_upload' ] );
+		add_filter( 'wp_generate_attachment_metadata', [ $this, 'wp_generate_attachment_metadata' ], 10, 3 );
+		add_filter( 'wp_check_filetype_and_ext', [ $this, 'wp_check_filetype_and_ext' ], 10, 5 );
+		add_filter( 'site_option_upload_filetypes', [ $this, 'filter_list_of_allowed_filetypes' ] );
+	}
+
+	/**
+	 * Helper function to check if svg uploads are already enabled.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @return bool
+	 */
+	private function svg_already_enabled() {
+		$allowed_mime_types = get_allowed_mime_types();
+		$mime_types         = array_values( $allowed_mime_types );
+
+		return in_array( self::MIME_TYPE, $mime_types, true );
+	}
+
+	/**
+	 * Enable SVG upload.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array $mime_types Mime types keyed by the file extension regex corresponding to those types.
+	 *
+	 * @return mixed
+	 */
+	public function upload_mimes_add_svg( array $mime_types ) {
+		// allow SVG file upload.
+		$mime_types['svg']  = self::MIME_TYPE;
+		$mime_types['svgz'] = self::MIME_TYPE;
+
+		return $mime_types;
+	}
+
+	/**
+	 * Adds SVG to list of mime types and file extensions
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param string[] $mime_types     Mime types keyed by the file extension regex
+	 *                                 corresponding to those types.
+	 *
+	 * @return array
+	 */
+	public function mime_types_add_svg( array $mime_types ) {
+		// allow SVG files.
+		$mime_types['svg'] = self::MIME_TYPE;
+
+		return $mime_types;
+	}
+
+	/**
+	 * Add SVG to allowed mime types.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array $mime_types Associative array of allowed mime types per media type (image, audio, video).
+	 *
+	 * @return array
+	 */
+	public function web_stories_allowed_mime_types( array $mime_types ) {
```

Comment:
Nit: there's no test for this method.

Also, should it avoid duplicate entries? Perhaps doing `array_unique( $mime_types['image'] )`

Reply:
I wasn't sure if there was value in testing a filter like this. But I can add a simple test.